Add nuance around stretched clusters (#77360)
elastic/elasticsearch: docs/reference/high-availability/cluster-design.asciidoc

[[high-availability-cluster-design-large-clusters]]
=== Resilience in larger clusters

It is not unusual for nodes to share some common infrastructure, such as network

> interconnects or a power supply. If so, you should plan for the failure of this

Author: I'm not sure about this. I mean it's correct but it does make the sentence much more complicated. Is it worth the extra words? Do we need to clarify that nodes on the same physical host share infrastructure like power and network? Seems kinda obvious to me but this is a genuine question, I'm not the one on the front line for this kind of thing.

infrastructure and ensure that such a failure would not affect too many of your
nodes. It is common practice to group all the nodes sharing some infrastructure
into _zones_ and to plan for the failure of any whole zone at once.

{es} expects its node-to-node connections to be reliable and have low latency

> and good bandwidth. Many of the tasks that {es} performs require multiple

Author: "Enough bandwidth" feels awkward to me, how about "adequate bandwidth"? See b0fae802ca7b566666e5043c2f0b3d375895301f.

round-trips between nodes. This means that a slow or unreliable interconnect
may have a significant effect on the performance and stability of your cluster.
A few milliseconds of latency added to each round-trip can quickly accumulate
into a noticeable performance penalty. {es} will automatically recover from a
network partition as quickly as it can but your cluster may be partly
unavailable during a partition and will need to spend time and resources to

> resynchronize any missing data and rebalance itself once a partition heals.

Author: Good point, thanks. Added a sentence at the end of this paragraph about recovery time in b0fae802ca7b566666e5043c2f0b3d375895301f.

If you have divided your cluster into zones then typically the network
connections within each zone are of higher quality than the connections between
the zones. You must make sure that the network connections between zones are of

> sufficiently high quality. You will see the best results by locating all your

Author: I'd rather be slightly more vague here: it's not just about performance, reliability is also a big deal.

zones within a single data center with each zone having its own independent
power supply and other supporting infrastructure. You can also _stretch_ your
cluster across nearby data centers as long as the network interconnection
between each pair of data centers is good enough.

[[high-availability-cluster-design-min-network-perf]] There is no specific
minimum network performance required to run a healthy {es} cluster. In theory a
cluster will work correctly even if the round-trip latency between nodes is
several hundred milliseconds. In practice if your network is that slow then the

> cluster performance will be very poor. In addition, slow networks are often

Reviewer: ```suggestion
cluster performance will likely be at unacceptable levels. In addition, slow networks are often
```

Author: I started with something like that but then I figured we'd have to dive into what "unacceptable" means and how you'd determine what is or isn't acceptable. I saw someone running a very stretched cluster over satellite links once. Its performance was terrible in an absolute sense, and yet it was still acceptable to them. There's certainly a place for that sort of discussion but it's not here.

[[high-availability-cluster-design-large-clusters]]
=== Resilience in larger clusters

It is not unusual for nodes to share some common infrastructure, such as network
interconnects or a power supply. If so, you should plan for the failure of this
infrastructure and ensure that such a failure would not affect too many of your
nodes. It is common practice to group all the nodes sharing some infrastructure
into _zones_ and to plan for the failure of any whole zone at once.

> {es} expects its node-to-node connections to be reliable and to have low
> latency and adequate bandwidth. Many of the tasks that {es} performs require
> multiple round-trips between nodes. This means that a slow or unreliable

Reviewer 2: ```suggestion
{es} expects node-to-node connections to be reliable, have low latency, and
have adequate bandwidth. Many {es} tasks require multiple round-trips between
nodes. A slow or unreliable
```

interconnect may have a significant effect on the performance and stability of

> your cluster. A few milliseconds of latency added to each round-trip can

Reviewer 2: I'd consider breaking the paragraph here.
```suggestion
your cluster.

For example, a few milliseconds of latency added to each round-trip can
```

> quickly accumulate into a noticeable performance penalty. {es} will

Reviewer 2: We may want to introduce network partitions and their cause after the first sentence.

```suggestion
quickly accumulate into a noticeable performance penalty. An unreliable network
may have frequent network partitions. {es} will
```

automatically recover from a network partition as quickly as it can but your
cluster may be partly unavailable during a partition and will need to spend
time and resources to resynchronize any missing data and rebalance itself once

> the partition heals. Recovering from a failure may involve copying a large

Reviewer 2: It seems like we're talking around reallocation and shard recovery here. Is there a reason we don't just directly mention and xref those two concepts?

Author: I didn't think we had any docs on those topics, at least not concept-level ones that would be suitable for linking from here. If you have some in mind then sure we can add links.

Reviewer 2: I think you're right. We have some setting reference:

- https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-cluster.html
- https://www.elastic.co/guide/en/elasticsearch/reference/master/recovery.html

However, I don't think those are great links to use here. I've opened https://github.com/elastic/elasticsearch/issues/77515 to track this gap and add those docs.

This looks fine to me in the meantime.

amount of data between nodes so the recovery time is often determined by the
available bandwidth.

If you have divided your cluster into zones then typically the network
connections within each zone are of higher quality than the connections between
the zones. You must make sure that the network connections between zones are of
sufficiently high quality. You will see the best results by locating all your
zones within a single data center with each zone having its own independent
power supply and other supporting infrastructure. You can also _stretch_ your
cluster across nearby data centers as long as the network interconnection
between each pair of data centers is good enough.

> [[high-availability-cluster-design-min-network-perf]] There is no specific

Reviewer 2: I'd add some whitespace for readability.
```suggestion
[[high-availability-cluster-design-min-network-perf]]
There is no specific
```

> minimum network performance required to run a healthy {es} cluster. In theory a

Reviewer 2: ```suggestion
minimum network performance required to run a healthy {es} cluster. In theory, a
```

cluster will work correctly even if the round-trip latency between nodes is

> several hundred milliseconds. In practice if your network is that slow then the

Reviewer 2: ```suggestion
several hundred milliseconds. In practice, if your network is that slow then the
```
